apps/spruce/src/pages/waterfall/utils.test.ts

```diff
@@ -137,6 +139,276 @@ describe("groupBuildVariants", () => {
   });
 });
 
+describe("caching functions", () => {
+  describe("mergeVersions", () => {
+    it("merges version arrays", () => {
+      const pagination = {
+        nextPageOrder: 0,
+        prevPageOrder: 0,
+        hasNextPage: true,
+        hasPrevPage: true,
+      };
+      expect(
+        mergeVersions(
+          {
+            flattenedVersions: versions.slice(0, 2),
+          },
+          {
+            flattenedVersions: versions.slice(2, -1),
+            pagination,
+          },
+          {
+            // @ts-expect-error
+            readField: (field, obj) => obj[field],
+          } as FieldFunctionOptions,
+        ),
+      ).toStrictEqual({ flattenedVersions: versions.slice(0, -1), pagination });
+    });
+
+    it("merges version when incoming is newer than existing", () => {
+      const pagination = {
+        nextPageOrder: 3,
+        prevPageOrder: 0,
+        hasNextPage: true,
+        hasPrevPage: false,
+      };
+      expect(
+        mergeVersions(
+          {
+            flattenedVersions: versions.slice(2, -1),
+          },
+          {
+            flattenedVersions: versions.slice(0, 2),
+            pagination,
+          },
+          {
+            // @ts-expect-error
+            readField: (field, obj) => obj[field],
+          } as FieldFunctionOptions,
+        ),
+      ).toStrictEqual({ flattenedVersions: versions.slice(0, -1), pagination });
+    });
+
+    it("deduplicates versions when merging", () => {
+      const pagination = {
+        nextPageOrder: 0,
+        prevPageOrder: 1,
+        hasNextPage: false,
+        hasPrevPage: true,
+      };
+      expect(
+        mergeVersions(
+          {
+            flattenedVersions: versions.slice(0, 4),
+          },
+          {
+            flattenedVersions: versions.slice(2),
+            pagination,
+          },
+          {
+            // @ts-expect-error
+            readField: (field, obj) => obj[field],
+          } as FieldFunctionOptions,
+        ),
+      ).toStrictEqual({ flattenedVersions: versions, pagination });
+    });
+
+    it("returns an identical cache when duplicate data is incoming", () => {
+      const pagination = {
+        nextPageOrder: 0,
+        prevPageOrder: 0,
+        hasNextPage: false,
+        hasPrevPage: false,
+      };
+      expect(
+        mergeVersions(
+          {
+            flattenedVersions: versions,
+          },
+          {
+            flattenedVersions: versions,
+            pagination,
+          },
+          {
+            // @ts-expect-error
+            readField: (field, obj) => obj[field],
+          } as FieldFunctionOptions,
+        ),
+      ).toStrictEqual({ flattenedVersions: versions, pagination });
+    });
+  });
+
```

Comment:
Can we add a test for leading inactive versions?

Reply:
Since `id: a` is inactive, I think this is covered by "reads the first page and returns limit active versions"? But lmk if you had something else in mind!